In [1]:
%autosave 0

Autosave disabled


In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv (r'C:\Users\Diego Cortínez\Desktop\Tarea 2\AB_NYC_2019.csv')

In [4]:
df = df[['latitude',
'longitude',
'price',
'minimum_nights',
'number_of_reviews',
'reviews_per_month',
'calculated_host_listings_count',
'availability_365']]

In [5]:
df

latitude  longitude  price  minimum_nights  number_of_reviews  \
0      40.64749  -73.97237    149               1                  9   
1      40.75362  -73.98377    225               1                 45   
2      40.80902  -73.94190    150               3                  0   
3      40.68514  -73.95976     89               1                270   
4      40.79851  -73.94399     80              10                  9   
...         ...        ...    ...             ...                ...   
48890  40.67853  -73.94995     70               2                  0   
48891  40.70184  -73.93317     40               4                  0   
48892  40.81475  -73.94867    115              10                  0   
48893  40.75751  -73.99112     55               1                  0   
48894  40.76404  -73.98933     90               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
2                    NaN                               1               365  
3                   4.64                               1               194  
4                   0.10                               1                 0  
...                  ...                             ...               ...  
48890                NaN                               2                 9  
48891                NaN                               2                36  
48892                NaN                               1                27  
48893                NaN                               6                 2  
48894                NaN                               1                23  

[48895 rows x 8 columns]

    Question 1: Find a feature with missing values. How many missing values does it have?

In [6]:
df.isnull().sum()

latitude                              0
longitude                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

    reviews_per_month                 10052

    Question 2: What's the median (50% percentile) for variable 'minimum_nights'?

In [7]:
df['minimum_nights'].median()

3.0

    median                3.0

    Split the data

    Shuffle

In [8]:
df

latitude  longitude  price  minimum_nights  number_of_reviews  \
0      40.64749  -73.97237    149               1                  9   
1      40.75362  -73.98377    225               1                 45   
2      40.80902  -73.94190    150               3                  0   
3      40.68514  -73.95976     89               1                270   
4      40.79851  -73.94399     80              10                  9   
...         ...        ...    ...             ...                ...   
48890  40.67853  -73.94995     70               2                  0   
48891  40.70184  -73.93317     40               4                  0   
48892  40.81475  -73.94867    115              10                  0   
48893  40.75751  -73.99112     55               1                  0   
48894  40.76404  -73.98933     90               7                  0   

       reviews_per_month  calculated_host_listings_count  availability_365  
0                   0.21                               6               365  
1                   0.38                               2               355  
2                    NaN                               1               365  
3                   4.64                               1               194  
4                   0.10                               1                 0  
...                  ...                             ...               ...  
48890                NaN                               2                 9  
48891                NaN                               2                36  
48892                NaN                               1                27  
48893                NaN                               6                 2  
48894                NaN                               1                23  

[48895 rows x 8 columns]

In [9]:
#For Q5
df_original = df.copy()

In [10]:
n = len(df)
idx = np.arange(n)
np.random.seed(42)
np.random.shuffle(idx)
idx


array([  879, 44383, 15394, ..., 38158,   860, 15795])

In [11]:
df = df.iloc[idx]
df

latitude  longitude  price  minimum_nights  number_of_reviews  \
879    40.64354  -73.97777     89               3                 62   
44383  40.70666  -73.90779     30              21                  0   
15394  40.76116  -73.99016    120               2                 17   
43230  40.70763  -74.01050    470               2                  5   
16332  40.79658  -73.93287    199               2                 30   
...         ...        ...    ...             ...                ...   
11284  40.84650  -73.94319     60               1                  0   
44732  40.73957  -74.00082     85               2                  4   
38158  40.78318  -73.97372    130              30                  1   
860    40.77508  -73.97990    150               2                 11   
15795  40.74657  -73.94555    120               5                  1   

       reviews_per_month  calculated_host_listings_count  availability_365  
879                 0.71                               1               189  
44383                NaN                               1                73  
15394               0.43                               1                 0  
43230               1.88                             327               272  
16332               0.80                               1                30  
...                  ...                             ...               ...  
11284                NaN                               1                 0  
44732               1.90                               1                76  
38158               0.34                               5               261  
860                 0.13                               1                 2  
15795               0.03                               1                 0  

[48895 rows x 8 columns]

    Split 

In [12]:
n_train = int(n * 0.6)
n_val = int(n * 0.2)
n_test  = n - n_val - n_train

    Make sure that the target value ('price') is not in your dataframe and Apply the log transformation to the price variable using the np.log1p() function.

In [13]:
df_train = df.iloc[idx[:n_train]]
df_val = df.iloc[idx[n_train:n_train+n_val]]
df_test = df.iloc[idx[n_train+n_val:]]

In [14]:
len(df_train), len(df_val), len(df_test)

(29337, 9779, 9779)

In [15]:
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [16]:
y_train = np.log1p(df_train.price.values)
y_val = np.log1p(df_val.price.values)
y_test = np.log1p(df_test.price.values)

In [17]:
del df_train['price']
del df_val['price']
del df_test['price']

In [18]:
df_train.head()

latitude  longitude  minimum_nights  number_of_reviews  reviews_per_month  \
0  40.70686  -73.91961              30                  7               0.16   
1  40.64917  -73.96236               3                  0                NaN   
2  40.69327  -73.95725               2                  0                NaN   
3  40.81372  -73.95612               1                 51               1.96   
4  40.71722  -73.87856              30                 33               0.31   

   calculated_host_listings_count  availability_365  
0                               1                 0  
1                               1                 0  
2                               1                87  
3                               1                67  
4                               5               240

    Question 3 
We need to deal with missing values for the column from Q1.
We have two options: fill it with 0 or with the mean of this variable.
Try both options. For each, train a linear regression model without regularization using the code from the lesssons.
For computing the mean, use the training only!
Use the validation dataset to evaluate the models and compare the RMSE of each option.
Round the RMSE scores to 2 decimal digits using round(score, 2)
Which option gives better RMSE?

In [19]:
def dot(xi, w):
    n = len(xi)
    
    res = 0.0
    
    for j in range(n):
        res = res + xi[j] * w[j]
    
    return res

In [20]:
def linear_regression(X):
    return X.dot(w_new)

In [21]:
def train_linear_regression(X, y):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [22]:
def prepare_X(df):
    X = df.values
    return X

In [23]:
def rmse(y, y_pred):
    se = (y - y_pred) ** 2
    mse = se.mean()
    return np.sqrt(mse)

In [24]:
df_train_zeros = df_train.copy()
df_train_mean = df_train.copy()
df_val_zeros = df_val.copy()
df_val_mean = df_val.copy()

In [25]:
df_train_zeros['reviews_per_month'] = df_train_zeros['reviews_per_month'].fillna(0)
df_train_mean['reviews_per_month'] = df_train_mean['reviews_per_month'].fillna(df_train['reviews_per_month'].mean())
df_val_zeros['reviews_per_month'] = df_val_zeros['reviews_per_month'].fillna(0)
df_val_mean['reviews_per_month'] = df_val_mean['reviews_per_month'].fillna(df_train['reviews_per_month'].mean())

In [26]:
df_train_zeros['reviews_per_month'].isnull().sum()

0

In [27]:
df_train_mean['reviews_per_month'].isnull().sum()

0

In [28]:
df_val_zeros['reviews_per_month'].isnull().sum()

0

In [29]:
df_val_mean['reviews_per_month'].isnull().sum()

0

In [30]:
X_train_zeros = prepare_X(df_train_zeros)
w0, w = train_linear_regression(X_train_zeros, y_train)

X_val_zeros = prepare_X(df_val_zeros)
y_pred = w0 + X_val_zeros.dot(w)
round(rmse(y_val, y_pred),2)

0.65

In [31]:
X_train_mean = prepare_X(df_train_mean)
w0, w = train_linear_regression(X_train_mean, y_train)

X_val_mean = prepare_X(df_val_mean)
y_pred = w0 + X_val_mean.dot(w)
round(rmse(y_val, y_pred),2)

0.65

    Question 4 
-Now let's train a regularized linear regression.
-For this question, fill the NAs with 0.
-Try different values of r from this list: [0, 0.000001, 0.0001, 0.001, 0.01, 0.1, 1, 5, 10].
-Use RMSE to evaluate the model on the validation dataset.
-Round the RMSE scores to 2 decimal digits.
-Which r gives the best RMSE?
-If there are multiple options, select the smallest r.

In [32]:
def train_linear_regression_reg(X, y, r=0.001):
    ones = np.ones(X.shape[0])
    X = np.column_stack([ones, X])

    XTX = X.T.dot(X)
    XTX = XTX + r * np.eye(XTX.shape[0])

    XTX_inv = np.linalg.inv(XTX)
    w_full = XTX_inv.dot(X.T).dot(y)
    
    return w_full[0], w_full[1:]

In [33]:
r = [0, 0.000001, 0.0001, 0.001, 0.01, 0.1, 1, 5, 10]

In [34]:
for value in r:
    X_train_zeros = prepare_X(df_train_zeros)
    w0, w = train_linear_regression_reg(X_train_zeros, y_train, r=value)

    X_val_zeros = prepare_X(df_val_zeros)
    y_pred = w0 + X_val_zeros.dot(w)
    round(rmse(y_val, y_pred),2)
    print(f"RMSE for r = {value}: {round(rmse(y_val, y_pred),2)}")



RMSE for r = 0: 0.65
RMSE for r = 1e-06: 0.65
RMSE for r = 0.0001: 0.65
RMSE for r = 0.001: 0.65
RMSE for r = 0.01: 0.66
RMSE for r = 0.1: 0.68
RMSE for r = 1: 0.68
RMSE for r = 5: 0.68
RMSE for r = 10: 0.68


    Question 5 
-We used seed 42 for splitting the data. Let's find out how selecting the seed influences our score.
-Try different seed values: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].
-For each seed, do the train/validation/test split with 60%/20%/20% distribution.
-Fill the missing values with 0 and train a model without regularization.
-For each seed, evaluate the model on the validation dataset and collect the RMSE scores.
-What's the standard deviation of all the scores? To compute the standard deviation, use np.std.
-Round the result to 3 decimal digits (round(std, 3))
-Note: Standard deviation shows how different the values are. If it's low, then all values are approximately the same. If it's high, the values are different. If standard deviation of scores is low, then our model is stable.

In [35]:
score_list = []
df = df_original
seed_values = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
for seed in seed_values:
    n = len(df)
    idx = np.arange(n)
    np.random.seed(seed)
    np.random.shuffle(idx)
    #split
    n_train = int(n * 0.6)
    n_val = int(n * 0.2)
    n_test  = n - n_val - n_train
    df_train = df.iloc[idx[:n_train]]
    df_val = df.iloc[idx[n_train:n_train+n_val]]
    df_test = df.iloc[idx[n_train+n_val:]]
    y_train = np.log1p(df_train.price.values)
    y_val = np.log1p(df_val.price.values)
    y_test = np.log1p(df_test.price.values)
    del df_train['price']
    del df_val['price']
    del df_test['price']
    X_train = df_train.values
    X_train = df_train.fillna(0).values
    X_val = df_val.fillna(0).values
    w0, w = train_linear_regression(X_train, y_train)
    y_pred = w0 + X_val.dot(w)
    print(f"RMSE for seed value : {seed} = {round(rmse(y_val, y_pred),2)}")
    score_list.append(rmse(y_val, y_pred))
    
print(f"STD of score : {round(np.std(score_list),3)}")

RMSE for seed value : 0 = 0.65
RMSE for seed value : 1 = 0.65
RMSE for seed value : 2 = 0.65
RMSE for seed value : 3 = 0.64
RMSE for seed value : 4 = 0.64
RMSE for seed value : 5 = 0.63
RMSE for seed value : 6 = 0.63
RMSE for seed value : 7 = 0.65
RMSE for seed value : 8 = 0.65
RMSE for seed value : 9 = 0.64
STD of score : 0.008


    Question 6 
-Split the dataset like previously, use seed 9.
-Combine train and validation datasets.
-Fill the missing values with 0 and train a model with r=0.001.
-What's the RMSE on the test dataset?

In [36]:
#Split datasets
n = len(df)
idx = np.arange(n)
np.random.seed(9)
np.random.shuffle(idx)
df = df_original.iloc[idx]
n_train = int(n * 0.6)
n_val = int(n * 0.2)
n_test  = n - n_val - n_train
df_train = df.iloc[idx[:n_train]]
df_val = df.iloc[idx[n_train:n_train+n_val]]
df_test = df.iloc[idx[n_train+n_val:]]
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)
y_train = np.log1p(df_train.price.values)
y_val = np.log1p(df_val.price.values)
y_test = np.log1p(df_test.price.values)
del df_train['price']
del df_val['price']
del df_test['price']
#Combine train and validation
df_combine = pd.concat([df_train,df_val])
df_combine = df_combine.reset_index(drop=True)
y_combine = np.concatenate((y_train, y_val))
#Filling missing values
df_combine_zeros = df_combine.copy()
df_test_zeros = df_test.copy()
df_combine_zeros['reviews_per_month'] = df_combine_zeros['reviews_per_month'].fillna(0)
df_test_zeros['reviews_per_month'] = df_test_zeros['reviews_per_month'].fillna(0)
#RMSE calculation
X_combine_zeros = prepare_X(df_combine_zeros)
w0, w = train_linear_regression_reg(X_combine_zeros, y_combine,r=0.001)

X_test_zeros = prepare_X(df_test_zeros)
y_pred = w0 + X_test_zeros.dot(w)
se = (y_test - y_pred) ** 2
mse = se.mean()
rmse = np.sqrt(mse)
    
print(f"RMSE value = {round(rmse,2)}")

RMSE value = 0.65
